### ISSUE TITLE
Nuitka error when using MSVC through prompt launchers

### ISSUE THREAD

**Issue by Keithcat1** [Nuitka/Nuitka#962]
I'm using Python 3.9, and MSVC on windows
I installed nuitka with:
pip install --upgrade nuitka
Then I activated the x64 native MSVC command prompt and tried to compile this code:
import code
code.interact()

But I got this:
C:\py>nuitka c.py
Nuitka:INFO: Starting Python compilation.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C level backend compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: cl (cl).
KeyError: 'WindowsSDKVersion':
  File "C:\python\lib\SITE-P~1\nuitka\build\BACKEN~1.SCO", line 401:
    scons_details_logger.info("Using Windows SDK %r." % env["WindowsSDKVersion"])
  File "C:\python\lib\SITE-P~1\nuitka\build\INLINE~1\lib\scons-3.1.2\SCons\Environment.py",
 line 405:
    return self._dict[key]

**Comment by kayhayen** [comment]
That is probably an indication of you not using the right kind of MSVC. This one doesn't have a Windows SDK, and therefore won't work. I can make the error nicer, but that not going to change much.

**Comment by kayhayen** [comment]
Also, don't activate any MSVC prompt, Nuitka finds the correct one, and if it does not, well it's not installed. Chances are you picking one that doesn't match your Python arch, etc.

I am including code that makes this key default to "None" in which case this should be output.

```
    if not env["WindowsSDKVersion"]:
        scons_logger.sysexit(
            "Error, the Windows SDK must be installed in Visual Studio."
        )
```

Specifically "build tools" are known to not contain the right thing. Not sure though how to recognize it.

**Comment by kayhayen** [comment]
You can see this on https://nuitka.net/doc/factory.html and it will be in the hotfix coming this weekend as of course, a `KeyError` is not a good user experience.

**Comment by kayhayen** [comment]
Ok, seeing that you had previous had a working setup, I got more suspicious, and it seems that the WindowsSDK version is in fact only set, when Scons detects it, but not otherwise. That is of course a pity, because in the future, we want to be able to depend on it, but there is the environment itself set from outside shell, so this is about repairing stuff.

**Comment by kayhayen** [comment]
So, factory was updated with this, sorry for my confusion, this seems important to fix.

**Comment by kayhayen** [comment]
Since you were talking about seconds to re-link in the other issue, one thing on my mind is that I saw it taking a second to run these batch scripts in scons, in case one is not running from the prompt.

Usually that's not too important, but I can see how it would be really nice for my CI it least, if that second could be taken away, esp. now that even `clcache` is non-forking, it's more likely to be closer to zero overhead. I just don't know how caching could work reliably, or how to make a cmd batch to run. Maybe you got more insights than me there @Keithcat1

**Comment by Keithcat1** [comment]
I just ran this from a normal shell with no build tools command prompt activated and it still gave me GCC:
C:\py>nuitka c.py --msvc 2019
Nuitka:INFO: Starting Python compilation.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C level backend compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: C:\Users\Keith\AppData\Local\Nuitka\Nuitka\gcc\x86_6
4\10.2.0-11.0.0-8.0.0-r5\mingw64\bin\gcc.exe (gcc.exe).
Nuitka-Scons:INFO: Compiled 11 C files using ccache.
Nuitka-Scons:INFO: Cached C files (using ccache) with result 'cache hit': 11
Nuitka:INFO: Keeping build directory 'c.build'.
Nuitka:INFO: Successfully created 'c.exe'.

BTW, could I give the --msvc option without an argument to have Nuitka automaticallyfind the right MSVC?

**Comment by kayhayen** [comment]
If you picked the wrong MSVC, it would deactivate it. However, I think what is happening with your option there, is that is not how it works, I bet it's more like 14.2 or 14.3 or something.

**Comment by kayhayen** [comment]
This is what it did for me:

```
scons: warning: VC version 2019.0 not installed.  C/C++ compilers are most likely not set correctly.
 Installed versions are: ['14.2', '14.1']
File "C:\Users\kayha\repos\Py2C\nuitka\build\SconsUtils.py", line 101, in createEnvironment

scons: warning: VC version 2019.0 not installed.  C/C++ compilers are most likely not set correctly.
 Installed versions are: ['14.2', '14.1']
File "C:\Users\kayha\repos\Py2C\nuitka\build\SconsUtils.py", line 101, in createEnvironment
Nuitka-Scons:INFO: Backend C compiler: C:\Users\kayha\AppData\Local\Nuitka\Nuitka\gcc\x86_64\10.2.0-11.0.0-8.0.0-r5\mingw64\bin\gcc.exe (gcc.exe).
```

That are 2019 and 2017 versions I think. Arguably I could add symbolic names for them to be accepted, it kind of makes sense, doesn't it, but then I don't really want to enter MS versioning messes.

However, you not getting that is probably due to having a pre-selected compiler from the prompt, and nothing is changing that, and honestly, I think it would be really hard to change it, as the one in PATH is probably simply going to win.

**Comment by kayhayen** [comment]
Noting the duplicate however, with me not using `--mingw` the MSVC detection is run twice. There is code to disable for `--mingw` that code from scons, that I have now moved to where it is running when the fallback compiler environment.

**Comment by kayhayen** [comment]
This code is suspicious: 

```
    # If we are on Windows, and MinGW is not enforced, lets see if we can
    # find "cl.exe", and if we do, disable automatic scan.
    if (
        os.name == "nt"
        and not mingw_mode
        and (
            getExecutablePath("cl", env=None) is not None
            or getExecutablePath("gcc", env=None) is not None
        )
    ):
        args["MSVC_USE_SCRIPT"] = False

```

I will have to checkout what `MSVC_USE_SCRIPT` would mean in your case. I guess, it might make the overriding from prompt work and the reason I disable it, is because there was a warning giving. This is very old stuff.

**Comment by Keithcat1** [comment]
How are you getting Nuitka to print that?
When I run this command, it still gives me GCC:
C:\py>nuitka --msvc=14.2 c.py                                                              
Nuitka:INFO: Starting Python compilation.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C level backend compilation via Scons.
Nuitka-Scons:INFO: Backend C compiler: C:\Users\Keith\AppData\Local\Nuitka\Nuitka\gcc\x86_6
4\10.2.0-11.0.0-8.0.0-r5\mingw64\bin\gcc.exe (gcc.exe).
Nuitka-Scons:INFO: Compiled 11 C files using ccache.
Nuitka-Scons:INFO: Cached C files (using ccache) with result 'cache hit': 11
Nuitka:INFO: Keeping build directory 'c.build'.
Nuitka:INFO: Successfully created 'c.exe'.
I upgraded my nuitka to the latest source and now it automatically uses CL from the visual studio prompt again.

**Comment by kayhayen** [comment]
@Keithcat1 your last comment is self-contradicting, is the last sentence the current state, i.e. what you describe is not happening anymore?

**Comment by kayhayen** [comment]
What I am thinking is that you might have this, when I use the wrong prompt, in this case native x64 with Python 32 bits.

```
C:\Users\kayha\repos\Py2C>c:\Python27_32\python.exe bin\nuitka tests\basics\Asserts.py
Nuitka:INFO: Starting Python compilation.
Nuitka:INFO: Completed Python level compilation and optimization.
Nuitka:INFO: Generating source code for C backend compiler.
Nuitka:INFO: Running data composer tool for optimal constant value handling.
Nuitka:INFO: Running C level backend compilation via Scons.
Nuitka-Scons:INFO: Mismatch between Python binary ('c:\\Python27_32\\python.exe' -> 'pei-i386') and C compiler (None -> 'pei-x86-64') arches, ignored!
Nuitka will use gcc from MinGW64 of winlibs to compile on Windows.
```

The mismatch is only a warning, because e.g. in the Github CI, there is always a compiler in the PATH, a 32 bit one, while the Python's used are 64 bits, so it's normal, albeit a gcc there. I wonder why it says `None` now, that's slightly ugly, but in principle that is the prime reason a `cl.exe` from a prompt would be ignored.

**Comment by kayhayen** [comment]
Ok, I will need to reverse logging and clearing of selected compiler, now it tells me this:

```
Nuitka-Scons:INFO: Mismatch between Python binary ('c:\\Python27_32\\python.exe' -> 'pei-i386') and C compiler ('C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX64\\x64\\cl.exe' -> 'pei-x86-64') arches, ignored!
```

Seems I picked a 2017 prompt there too.

**Comment by kayhayen** [comment]
And now, with --msvc=14.2 I can make it switch away from the wrong 14.1 of the prompt. Something that is maybe a bit uninformative as that it says only `cl` and not the full path, which I think I did in other places, but I don't want to immediately go there.

**Comment by kayhayen** [comment]
I am also making a check if you specified the MSVC version for the fallback to be used. Using it would be against what you said.

**Comment by kayhayen** [comment]
So, on factory it should now work to override a selected prompt with --msvc-version. The error outputs should contain compiler paths rather than None (a gcc too old error had the same issue) and there is no fallback to gcc being done when --msvc-version is specified. In a way, that is a way of enforcing MSVC now. 

Please try it out, and let me know if that works as expected for you.

**Comment by kayhayen** [comment]
While these other changes are not, the fix for build tools to work was in 0.6.11.4 and I think that should work now, so I am closing this, these other things are just general improvements.
